frontend/src/framework/GuiMessageBroker.ts

```diff
@@ -0,0 +1,179 @@
+import React from "react";
+
+import { Point } from "@lib/utils/geometry";
+
+export enum DrawerContent {
+    ModuleSettings = "ModuleSettings",
+    ModulesList = "ModulesList",
+    TemplatesList = "TemplatesList",
+    SyncSettings = "SyncSettings",
+    ColorPaletteSettings = "ColorPaletteSettings",
+}
+
+export enum GuiState {
+    DrawerContent = "drawerContent",
+    SettingsPanelWidthInPercent = "settingsPanelWidthInPercent",
+    LoadingEnsembleSet = "loadingEnsembleSet",
+    ActiveModuleInstanceId = "activeModuleInstanceId",
+}
+
+export enum GuiEvent {
+    ModuleHeaderPointerDown = "moduleHeaderPointerDown",
+    NewModulePointerDown = "newModulePointerDown",
+    RemoveModuleInstanceRequest = "removeModuleInstanceRequest",
+}
+
+export type GuiEventPayloads = {
+    [GuiEvent.ModuleHeaderPointerDown]: {
+        moduleInstanceId: string;
+        elementPosition: Point;
+        pointerPosition: Point;
+    };
+    [GuiEvent.NewModulePointerDown]: {
+        moduleName: string;
+        elementPosition: Point;
+        pointerPosition: Point;
+    };
+    [GuiEvent.RemoveModuleInstanceRequest]: {
+        moduleInstanceId: string;
+    };
+};
+
+type GuiStateTypes = {
+    [GuiState.DrawerContent]: DrawerContent;
+    [GuiState.SettingsPanelWidthInPercent]: number;
+    [GuiState.LoadingEnsembleSet]: boolean;
+    [GuiState.ActiveModuleInstanceId]: string;
+};
+
+const defaultStates: Map<GuiState, any> = new Map();
+defaultStates.set(GuiState.DrawerContent, DrawerContent.ModuleSettings);
+defaultStates.set(GuiState.SettingsPanelWidthInPercent, 30);
+defaultStates.set(GuiState.LoadingEnsembleSet, false);
+defaultStates.set(GuiState.ActiveModuleInstanceId, "");
+
+const persistentStates: GuiState[] = [GuiState.SettingsPanelWidthInPercent];
+
+export class GuiMessageBroker {
+    private _eventListeners: Map<GuiEvent, Set<(event: any) => void>>;
+    private _stateSubscribers: Map<GuiState, Set<(state: any) => void>>;
+    private _storedValues: Map<GuiState, any>;
+
+    constructor() {
+        this._eventListeners = new Map();
+        this._stateSubscribers = new Map();
+        this._storedValues = defaultStates;
+
+        this.loadPersistentStates();
+    }
+
+    private loadPersistentStates() {
+        persistentStates.forEach((state) => {
+            const value = localStorage.getItem(state);
+            if (value) {
+                this._storedValues.set(state, JSON.parse(value));
+            }
+        });
+    }
+
+    private maybeSavePersistentState(state: GuiState) {
+        if (persistentStates.includes(state)) {
+            // For now, persistent states are only stored in localStorage
+            // However, in the future, we may want to store at least some of them in a database on the server
+            localStorage.setItem(state, JSON.stringify(this._storedValues.get(state)));
+        }
+    }
+
+    subscribeToEvent<K extends GuiEvent>(event: K, callback: (event: GuiEventPayloads[K]) => void) {
```

Comment:
Rename `event` to `payload` or `eventPayload`

Any particular reason for using `K` instead of `T` as we usually do? This goes for many of the member functions.
Maybe we should adopt another scheme for naming these parametric types? Noticed that React Query uses names such as `TQueryFnData` and `TError`